text/0201-refactoring-support.md

```diff
@@ -0,0 +1,193 @@
+
+With this refactoring support you can organize your code without causing impact to your existing stacks.
+
+## Internal FAQ
+
+### Why are we doing this?
+
+The ability to reorganize your code to make it more readable is fundamental to good software engineering, which CDK aims to support. Its current behavior is hindering good software practices.
+
+### Why should we _not_ do this?
+
+The current product has some affordances that attempt to give control over logical IDs to users, but those solutions are incomplete and do not work at scale.
+
+### Currently available solutions
+
+Currently, the following mechanism are available to users:
+
+- `CfnElement.overrideLogicalId(id)`/`Stack.allocateLogicalId(element, id)`
+- `Stack.renameLogicalId(oldId, newId)`
+
+The first two are equivalent: given an (L1) resource, they allow controlling the logical ID of that resource in the containing stack.
+
+The second allows for renaming a new logical id (after refactoring) to a known old logical ID (from before the refactoring), in order to prevent resource replacement.
+
+Downsides of the existing mechanisms are:
+
+- **Non-locality**: for many reasons[1], the responsibility of applying logical ID control must be on the person who assembles the `Stack`.
+  - In a complex application with a deeply layered construct tree, the person who maintains the Stack may be far removed (in time and knowledge) from the person who actually performed the refactoring.
+  - If a construct contains a refactoring and is used in multiple Stacks, all Stacks need to have a refactoring applied, which does not scale.
+- **Logical ids only**: current methods only allow control of the logical ID. However, the construct library contains other (physical) identifiers that are derived from the construct tree, which can currently not be controlled.
+
+In contrast, the mechanism proposed here requires adding a bit of local information at the time and place where the refactoring is performed, and will affect all identifiers derived from the construct tree.
+
+### What is the technical solution (design) of this feature?
+
+The feature will be implemented in the `constructs` library. A record of refactorings will be maintained in each construct node, and construct nodes will have two methods to calculate a construct path:
+
+- The "real" construct path
+- The "identifier" construct path
+
+All methods that calculate identifiers will be rewritten to use the second root path.
+
+We will add misuse protection. It is important that when a construct is "virtually" moved into the location of another, that that location is not reused by a meaningful construct, as their identifiers would start to collide. In the example given above:
+
+```ts
+this.node.refactor(this, 'MyBucket', 'MyConstruct/MyBucket');
```

Comment:
Since the API is a method call and not "new"-ing something, it reads to me like an operation, not a declaration, so as a user I'd expect I could call this multiple times. Would something like this be possible?

```ts
this.node.refactor(this, 'MyBucket', 'MyConstructV1/MyBucket');
this.node.refactor(this, 'MyConstructV1/MyBucket', 'MyConstructV2/MyBucket');
```